[WPE] WPE Platform: add wpe_input_method_context_filter_key_event

zhani opened the pull request:
#### f3070e7128adc401c89b6b7fc843194dffb7c3d8
<pre>
[WPE] WPE Platform: add wpe_input_method_context_filter_key_event
<a href="https://bugs.webkit.org/show_bug.cgi?id=275893">https://bugs.webkit.org/show_bug.cgi?id=275893</a>

Reviewed by Carlos Garcia Campos.

Add missing wpe_input_method_context_filter_key_event to
WPEPlatform

* Source/WebKit/UIProcess/API/glib/InputMethodFilter.h:
(WebKit::InputMethodFilter::setUseWPEPlatformEvents):
* Source/WebKit/UIProcess/API/glib/WebKitInputMethodContext.h.in:
* Source/WebKit/UIProcess/API/wpe/InputMethodFilterWPE.cpp:
(WebKit::InputMethodFilter::platformEventKeyIsKeyPress const):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::m_backend):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
(webkitInputMethodContextImplWPEFilterKeyEvent):
(webkit_input_method_context_impl_wpe_class_init):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextWPE.cpp:
(webkit_input_method_context_filter_key_event):
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
(wpe_input_method_context_filter_key_event):
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestInputMethodContext.cpp:
(webkitInputMethodContextMockFilterKeyEvent):

Canonical link: <a href="https://commits.webkit.org/280448@main">https://commits.webkit.org/280448@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b31aba3fe66b80060c2343839bfb2593c56271d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56676 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36003 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9149 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60290 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7117 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58803 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7308 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4986 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58706 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33843 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48916 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26770 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30622 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6247 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6120 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52565 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6519 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61973 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/588 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6620 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/590 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48974 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53184 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/507 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8427 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31831 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32917 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34001 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32663 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/InputMethodFilter.h:
```diff
@@ -125,6 +129,10 @@ class InputMethodFilter {
         uint64_t selectionPosition;
     } m_surrounding;
+
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM)
```
```suggestion
#if ENABLE(WPE_PLATFORM)
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitInputMethodContext.h.in:
```diff
@@ -126,7 +126,7 @@ struct _WebKitInputMethodContextClass {
 #elif PLATFORM(WPE)
     gboolean (* filter_key_event)   (WebKitInputMethodContext        *context,
-                                     struct wpe_input_keyboard_event *key_event);
+                                     void                            *key_event);
```
```suggestion
                                     gpointer                         key_event);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitInputMethodContext.h.in:
```diff
@@ -184,7 +184,7 @@ webkit_input_method_context_filter_key_event   (WebKitInputMethodContext   *cont
 WEBKIT_API gboolean
 webkit_input_method_context_filter_key_event   (WebKitInputMethodContext        *context,
-                                                struct wpe_input_keyboard_event *key_event);
+                                                void                            *key_event);
```
```suggestion
                                                gpointer                         key_event);
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/InputMethodFilterWPE.cpp:
```diff
@@ -34,7 +38,14 @@ IntRect InputMethodFilter::platformTransformCursorRectToViewCoordinates(const In
-    return event->pressed;
+#if ENABLE(WPE_PLATFORM)
+    if (m_useWPEPlatformEvents) {
+        auto* wpe_platform_event = static_cast<WPEEvent*>(event);
+        return wpe_event_get_event_type(wpe_platform_event) == WPE_EVENT_KEYBOARD_KEY_DOWN;
```
```suggestion
        return wpe_event_get_event_type(WPE_EVENT(event)) == WPE_EVENT_KEYBOARD_KEY_DOWN;
```

zhani replied:
WPEvent is not gobject but boxed type

carlosgcampos replied:
Right, sorry I forgot it

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -177,15 +178,20 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
+                auto filterResult = webView.m_inputMethodFilter.filterKeyEvent(event);
+                if (filterResult.handled)
+                    return filterResult.handled;
                 webView.page().handleKeyboardEvent(WebKit::NativeWebKeyboardEvent(event, String(), false));
```
```suggestion
                if (!filterResult.handled)
                    webView.page().handleKeyboardEvent(WebKit::NativeWebKeyboardEvent(event, String(), false));
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -177,15 +178,20 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
+                auto filterResult = webView.m_inputMethodFilter.filterKeyEvent(event);
+                if (filterResult.handled)
+                    return filterResult.handled;
                 webView.page().handleKeyboardEvent(WebKit::NativeWebKeyboardEvent(event, String(), webView.m_keyAutoRepeatHandler.keyPress(wpe_event_keyboard_get_keycode(event))));
```
```suggestion
                if (!filterResult.handled)
                    webView.page().handleKeyboardEvent(WebKit::NativeWebKeyboardEvent(event, String(), webView.m_keyAutoRepeatHandler.keyPress(wpe_event_keyboard_get_keycode(event))));
```

carlosgcampos replied:
You have to pass `filterResult.keyText` instead of `String()`

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -138,6 +138,13 @@ static void webkitInputMethodContextImplWPEGetPreedit(WebKitInputMethodContext*
+static gboolean webkitInputMethodContextImplWPEFilterKeyEvent(WebKitInputMethodContext* context, void* keyEvent)
+{
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    auto* wpeEvent = static_cast<WPEEvent*>(keyEvent);
+    return wpe_input_method_context_filter_key_event(priv->context.get(), wpeEvent);
```
```suggestion
    return wpe_input_method_context_filter_key_event(priv->context.get(), WPE_EVENT(keyEvent));
```

zhani replied:
WPEvent is not gobject but boxed type

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -410,6 +410,28 @@ void wpe_input_method_context_get_preedit_string(WPEInputMethodContext* context,
+ * should be done for this key event.
+ *
+ * Returns: %TRUE if the input method handled the key event.
+ *
```
```suggestion
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -410,6 +410,28 @@ void wpe_input_method_context_get_preedit_string(WPEInputMethodContext* context,
+    auto* imeClass = WPE_INPUT_METHOD_CONTEXT_GET_CLASS(context);
+    if (imeClass->filter_key_event)
+        return imeClass->filter_key_event(context, event);
+    return false;
```
```suggestion
    return FALSE;
```

zhani replied:
Fixed

zhani committed:
[WPE] WPE Platform: add wpe_input_method_context_filter_key_event
https://bugs.webkit.org/show_bug.cgi?id=275893

Reviewed by Carlos Garcia Campos.

Add missing wpe_input_method_context_filter_key_event to
WPEPlatform

* Source/WebKit/UIProcess/API/glib/InputMethodFilter.h:
(WebKit::InputMethodFilter::setUseWPEPlatformEvents):
* Source/WebKit/UIProcess/API/glib/WebKitInputMethodContext.h.in:
* Source/WebKit/UIProcess/API/wpe/InputMethodFilterWPE.cpp:
(WebKit::InputMethodFilter::platformEventKeyIsKeyPress const):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::m_backend):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
(webkitInputMethodContextImplWPEFilterKeyEvent):
(webkit_input_method_context_impl_wpe_class_init):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextWPE.cpp:
(webkit_input_method_context_filter_key_event):
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
(wpe_input_method_context_filter_key_event):
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestInputMethodContext.cpp:
(webkitInputMethodContextMockFilterKeyEvent):

Canonical link: https://commits.webkit.org/280448@main
